django/apps/registry.py

```diff
@@ -344,6 +346,79 @@ def clear_cache(self):
             for model in self.get_models(include_auto_created=True):
                 model._meta._expire_cache()
 
+    @staticmethod
+    def _model_key(model_or_name):
+        """
+        Returns an (app_label, model_name) tuple from a model or string.
+        """
+        if isinstance(model_or_name, six.string_types):
+            app_label, model_name = model_or_name.split(".")
+        else:
+            # It's actually a model class.
+            app_label = model_or_name._meta.app_label
+            model_name = model_or_name._meta.object_name
+        return app_label, model_name
+
+    def lazy_model_operation(self, function, *models_or_names, **kwargs):
+        """
+        Take a function and a number of models or model names, and when all
+        the corresponding models have been loaded, call the function with
+        the model classes and any optional as its arguments.
+
+        Each element of models_or names may be
+            * an actual model class
+            * a string of the form "app_label.ModelName"
+            * a tuple of strings of the form (app_label, model_name)
+
+        The function passed to this method must accept n positional arguments,
+        where n=len(models_or_names), plus whatever keyword arguments you pass
+        in kwargs.
+        """
+
+        if not self._class_prepared_handler_registered:
+            from django.db.models import signals
+            signals.class_prepared.connect(self.do_pending_lookups)
+            self._class_prepared_handler_registered = True
+
+        # Eagerly parse all model strings so we can fail immediately
+        # if any are plainly invalid.
+        model_keys = [self._model_key(m) if not isinstance(m, tuple) else m
+                      for m in models_or_names]
+
+        # If this function depends on more than one model, recursively call add
+        # for each, partially applying the given function on each iteration.
+        model_key, more_models = model_keys[0], model_keys[1:]
+        if more_models:
+            inner_function = function
+            function = lambda model, **kwargs: (
```

Comment:
I'll try rewriting this bit as you suggest.

> does Django itself actually need [multiple model] capability anywhere?

Presently, in this patch, every call to ``lazy_related_operation`` waits for the "local" model as well as each related model passed in. That means if it's called during class construction, even if the related models are already registered, the operation won't actually happen until the local model is fully prepared and registered as well. I feel that's cleaner and more consistent – it gives you more certainty about the state of the both models when the operation is performed.

Beyond that, this patch waits for two related models in [``create_many_to_many_intermediary_model``](https://github.com/django/django/pull/3984/files?diff=unified#diff-3010fc5a498b7171c342520f34507968R2028) – the yet-to-be-created through model, and the model on the other side of the relation. This could be avoided, but I think it makes the API much nicer at the cost of a little well-defined internal trickiness.